pdp-info/info.go

```diff
@@ -0,0 +1,150 @@
+package info
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type streamObj struct {
+	stream PipService_GetRequestsStreamServer
+	err    chan error
+}
+
+type streamsPool struct {
+	streams map[string][]streamObj
+	sync.RWMutex
+}
+
+func newStreamsPool() *streamsPool {
+	return &streamsPool{
+		streams: make(map[string][]streamObj),
+	}
+}
+
+func (p *streamsPool) RegisterStream(stream PipService_GetRequestsStreamServer) chan error {
+	cerr := make(chan error, 1)
+
+	resp, err := stream.Recv()
+	if err != nil {
+		cerr <- err
+		return cerr
+	}
+
+	inAttrs := resp.GetAttributes()
+	if len(inAttrs) != 1 {
+		cerr <- fmt.Errorf("Expected exactly one attribute, got: %d", len(inAttrs))
+		return cerr
+	}
+
+	name := inAttrs[0].GetValue()
+	s := streamObj{stream: stream, err: cerr}
+
+	p.Lock()
+	defer p.Unlock()
+
+	if _, ok := p.streams[name]; !ok {
+		p.streams[name] = []streamObj{s}
+	} else {
+		p.streams[name] = append(p.streams[name], s)
+	}
+
+	log.Infof("New PIP [%s] stream registered", name)
+
+	return cerr
+}
+
+func (p *streamsPool) GetStream(name string) (stream PipService_GetRequestsStreamServer,
+	index int, err error) {
+
+	p.RLock()
+	defer p.RUnlock()
+
+	list := p.streams[name]
+	if len(list) == 0 {
+		err = fmt.Errorf("PIP [%s] streams not registered", name)
+		log.Errorf(err.Error())
+		return
+	}
+
+	// streams random balancer
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	index = r.Intn(len(list))
+	stream = list[index].stream
+	return
+}
+
+func (p *streamsPool) DeleteStream(name string, index int) {
+	p.Lock()
+	defer p.Unlock()
+
+	if list, ok := p.streams[name]; ok {
+		if index >= 0 && index < len(list) {
+			err := fmt.Errorf("PIP [%s] stream deleted", name)
+			list[index].err <- err
+			p.streams[name] = append(list[:index], list[index+1:]...)
+			log.Errorf(err.Error())
+		}
+	}
+}
+
+type PipService interface {
+	Get(uri string, attrs []*Attribute) ([]*Attribute, error)
+	PipServiceServer
+}
+
+type pipService struct {
+	pool *streamsPool
+}
+
+func (p *pipService) Get(uri string, attrs []*Attribute) ([]*Attribute, error) {
+	var service, query string
+	if i := strings.Index(uri, "/"); i >= 0 {
+		service = uri[:i]
+		query = uri[i+1:]
+	} else {
+		service = uri
+	}
+
+	var streamErr error
+	for {
+		stream, index, err := p.pool.GetStream(service)
+		if err != nil {
+			streamErr = err
+			break
+		}
+
+		req := &Request{Query: query, Attributes: attrs}
+		if err := stream.Send(req); err != nil {
+			p.pool.DeleteStream(service, index)
```

Comment:
at this point index may be invalid, since we unlocked the map mutex

Reply:
Good catch - I am going to use map instead list for fix this